tests/unit/test_contracts/bls_primitives_tests.cpp

```diff
@@ -0,0 +1,361 @@
+        }
+
+        for(uint8_t k = 16 ; k < 64; k++) {
+            // Remaining 48 blocks
+            m[k] = sha256::sig1(m[k - 2]) + m[k - 7] + sha256::sig0(m[k - 15]) + m[k - 16];
+        }
+
+        for(uint8_t i = 0 ; i < 8 ; i++) {
+            state[i] = m_state[i];
+        }
+
+        for(uint8_t i = 0; i < 64; i++) {
+            maj   = sha256::majority(state[0], state[1], state[2]);
+            xorA  = sha256::rotr(state[0], 2) ^ sha256::rotr(state[0], 13) ^ sha256::rotr(state[0], 22);
+
+            ch = choose(state[4], state[5], state[6]);
+
+            xorE  = sha256::rotr(state[4], 6) ^ sha256::rotr(state[4], 11) ^ sha256::rotr(state[4], 25);
+
+            sum  = m[i] + K[i] + state[7] + ch + xorE;
+            newA = xorA + maj + sum;
+            newE = state[3] + sum;
+
+            state[7] = state[6];
+            state[6] = state[5];
+            state[5] = state[4];
+            state[4] = newE;
+            state[3] = state[2];
+            state[2] = state[1];
+            state[1] = state[0];
+            state[0] = newA;
+        }
+
+        for(uint8_t i = 0 ; i < 8 ; i++) {
+            m_state[i] += state[i];
+        }
+    }
+    void pad() {
+        uint64_t i = m_blocklen;
+        uint8_t end = m_blocklen < 56 ? 56 : 64;
+
+        m_data[i++] = 0x80; // Append a bit 1
+        while(i < end) {
+            m_data[i++] = 0x00; // Pad with zeros
+        }
+
+        if(m_blocklen >= 56) {
+            transform();
+            memset(m_data, 0, 56);
+        }
+
+        // Append to the padding the total message's length in bits and transform.
+        m_bitlen += m_blocklen * 8;
+        m_data[63] = m_bitlen;
+        m_data[62] = m_bitlen >> 8;
+        m_data[61] = m_bitlen >> 16;
+        m_data[60] = m_bitlen >> 24;
+        m_data[59] = m_bitlen >> 32;
+        m_data[58] = m_bitlen >> 40;
+        m_data[57] = m_bitlen >> 48;
+        m_data[56] = m_bitlen >> 56;
+        transform();
+    }
+    void revert(std::array<uint8_t, 32>& hash) {
+        // SHA uses big endian byte ordering
+        // Revert all bytes
+        for(uint8_t i = 0 ; i < 4 ; i++) {
+            for(uint8_t j = 0 ; j < 8 ; j++) {
+                hash[i + (j * 4)] = (m_state[j] >> (24 - i * 8)) & 0x000000ff;
+            }
+        }
+    }
+};
+} // namespace bls12_381
+
+class [[eosio::contract]] bls_primitives_tests : public contract{
+    public:
+        using contract::contract;
+
+        [[eosio::action]]
+        void testg1add(const std::vector<char>& op1, const std::vector<char>& op2, const std::vector<char>& res) {
+            check(op1.size() == std::tuple_size<bls_g1>::value, "wrong op1 size passed");
+            check(op2.size() == std::tuple_size<bls_g1>::value, "wrong op2 size passed");
+            bls_g1 r;
+            bls_g1_add(reinterpret_cast<const bls_g1&>(op1[0]), reinterpret_cast<const bls_g1&>(op2[0]), r);
```

Comment:
Huh, is that code from me? If so, I totally agree with your change. Can't believe I wrote that.. seems very wrong to me as well!

Reply:
other than personal preference I do not know any issues with that but I'll change that if that looks too unusual

Reply:
Sorry if that was a change from you. I didn't mean to be rude or something, just thought I committed this somehow by accident because it looks indeed very unusual to me. No offense!